### PR TITLE
fix: kl_div and superset for torch backend

### PR DESCRIPTION
**Pull request by ShreyanshBardia**
The previous implementation of `kl_div` was buggy and didn't follow input style of native backends i.e. `torch` and `paddle`. 
This PR aims at adding a mixed implementation of `kl_div`, earlier redundant code was present in backends (`numpy`, `jax`, `tensorflow`) which has been removed in favour of a compositional implementation added at `ivy/experimental/losses.py`. The previous implementation also didn't include a parameter `log_target` which is present in `torch` backend, this has been added following `Superset Approach`

### CODE REVIEW

**Comment by ivy-leaves** [comment]
If you are working on an open task, please **edit** the PR description to link to the issue you've created.

For more information, please check [ToDo List Issues Guide](https://unify.ai/docs/ivy/overview/contributing/the_basics.html#todo-list-issues).

Thank you :hugs:

File: ivy/data_classes/array/experimental/losses.py [hunk 1]
```diff
@@ -258,7 +258,7 @@ def kl_div(
         /,
         *,
         reduction: Optional[str] = "mean",
-        out: Optional[ivy.Array] = None,
+        log_target=False,
```
Comment (HaiderSultanArc) [inline ivy/data_classes/array/experimental/losses.py:261]: I don't know about this change 🤔 why did you remove the `out` parameter and introduce a new parameter in the functional API of ivy?

Comment (HaiderSultanArc) [reply ivy/data_classes/array/experimental/losses.py:261]: You can not remove `out` parameter. It's part of the standard for API that we are following all across the Functional API

File: ivy/data_classes/array/experimental/losses.py [hunk 2]
```diff
@@ -294,7 +294,9 @@ def kl_div(
         >>> print(output_array)
         ivy.array(0.0916)
         """
-        return ivy.kl_div(self._data, target, reduction=reduction, out=out)
```
Comment (HaiderSultanArc) [inline ivy/data_classes/array/experimental/losses.py:297]: Same here, we are not supposed to remove `out` parameter from Functional API! 😄

File: ivy/functional/backends/jax/experimental/losses.py [hunk 3]
```diff
@@ -64,26 +64,6 @@ def soft_margin_loss(
         return loss
 
 
-def kl_div(
```
Comment (HaiderSultanArc) [inline ivy/functional/backends/jax/experimental/losses.py:67]: You just removed the backend implementation from jax? 🤔 I am not following what you are trying to do. Can you elaborate more by providing description in the PR?

File: ivy/functional/backends/numpy/experimental/losses.py [hunk 4]
```diff
@@ -75,29 +75,6 @@ def soft_margin_loss(
         return loss
 
 
-@with_unsupported_dtypes({"1.26.0 and below": ("bool", "bfloat16")}, backend_version)
```
Comment (HaiderSultanArc) [inline ivy/functional/backends/numpy/experimental/losses.py:78]: Same here. You can't remove backend implementation from any backend unless you are creating a new implementation that is better

File: ivy/functional/ivy/experimental/losses.py [hunk 5]
```diff
@@ -421,7 +421,7 @@ def kl_div(
     /,
     *,
     reduction: Optional[str] = "mean",
-    out: Optional[ivy.Array] = None,
```
Comment (HaiderSultanArc) [inline ivy/functional/ivy/experimental/losses.py:424]: Again, we can't just introduce and remove parameters from the Functional API of ivy without any explaination.

Comment (ShreyanshBardia) [reply ivy/functional/ivy/experimental/losses.py:424]: Is it necessary for every function to have `out` parameter. I was not aware of that, will make the required changes

File: ivy/functional/backends/jax/experimental/losses.py (hunk 3 above)
Comment (ShreyanshBardia) [reply ivy/functional/backends/jax/experimental/losses.py:67]: We don't need a backend specific in these cases because there is no efficient implementation of it, so even if we define a compositional implementation, the function will fall back on to that (the one defined in `ivy/experimental/losses.py`). If you'll take a look at the previous implementations, they were using a compositional approach only, there is no primary function that exists for `numpy`, `jax`, `tensorflow`

Comment (ShreyanshBardia) [reply ivy/functional/backends/jax/experimental/losses.py:67]: We can define a compositional implementation in `ivy/experimental/losses.py`, so that for all the backends not having a primary or efficient implementation, this compositional implementation is called and we don't have to write the same function in each backend

File: ivy/data_classes/array/experimental/losses.py (hunk 1 above)
Comment (ShreyanshBardia) [reply ivy/data_classes/array/experimental/losses.py:261]: I'll add back the `out` parameter, but as for the `log_target`, aren't we trying to capture the superset behaviour? If so then I think it is important to include it, given that `torch` implements it

**Comment by ShreyanshBardia** [comment]
@HaiderSultanArc I have added back the mistakenly removed `out` argument, also I have updated the PR description to be more elaborate. Can you please take a look now? (Apologies for being ignorant about the `out` param) Thanks!

File: ivy/functional/backends/jax/experimental/losses.py (hunk 3 above)
Comment (HaiderSultanArc) [reply ivy/functional/backends/jax/experimental/losses.py:67]: We still need to implement backend-specific implementation 😅 . Even if a backend doesn't provide the implementation we still provide that (as the composition of other ivy functions) because other backends are providing it.

File: ivy/functional/ivy/experimental/losses.py (hunk 5 above)
Comment (HaiderSultanArc) [reply ivy/functional/ivy/experimental/losses.py:424]: I'll have a look at `poisson_nll_loss`

Comment (HaiderSultanArc) [reply ivy/functional/ivy/experimental/losses.py:424]: But yes it is necessary

File: ivy/functional/backends/jax/experimental/losses.py (hunk 3 above)
Comment (HaiderSultanArc) [reply ivy/functional/backends/jax/experimental/losses.py:67]: Got it! So only Torch has an implementation of k1_div in its functional API. 😄 Looks good to me.

File: ivy_tests/test_ivy/test_functional/test_experimental/test_nn/test_losses.py [hunk 6]
```diff
@@ -77,20 +78,23 @@ def test_huber_loss(
...
     dtype_and_target,
     reduction,
+    log_target,
     test_flags,
     backend_fw,
     fn_name,
     on_device,
 ):
     input_dtype, input = dtype_and_input
+    input[0] = np.log(input[0])
```
Comment (HaiderSultanArc) [inline ivy_tests/test_ivy/test_functional/test_experimental/test_nn/test_losses.py:94]: Why are you taking log of the input here? 🤔

File: ivy_tests/test_ivy/test_functional/test_experimental/test_nn/test_losses.py [hunk 7]
```diff
@@ -77,20 +78,23 @@ def test_huber_loss(
...
     target_dtype, target = dtype_and_target
-
+    if log_target:
+        target[0] = np.log(target[0])
```
Comment (HaiderSultanArc) [inline ivy_tests/test_ivy/test_functional/test_experimental/test_nn/test_losses.py:97]: if `log_target` is true then the user is expected to input a log of their target? 🤔

Comment (ShreyanshBardia) [reply ivy_tests/test_ivy/test_functional/test_experimental/test_nn/test_losses.py:97]: Yes that's correct. This is again from [`torch`](https://pytorch.org/docs/stable/generated/torch.nn.functional.kl_div.html) documentation  `A flag indicating whether target is passed in the log space. It is recommended to pass certain distributions (like softmax) in the log space to avoid numerical issues caused by explicit log. Default: False`

**Comment by HaiderSultanArc** [comment]
@ShreyanshBardia merge main into your branch and let the CI run again.

**Comment by ShreyanshBardia** [comment]
@HaiderSultanArc not sure about the errors on CI, they seem to run find locally

**Comment by HaiderSultanArc** [comment]
@ShreyanshBardia are all the tests passing locally? 🤔

**Comment by HaiderSultanArc** [comment]
Let's just ask for a final review from @vedpatwardhan 😄

File: ivy/functional/backends/torch/experimental/losses.py [hunk 8]
```diff
@@ -102,18 +103,8 @@ def soft_margin_loss(
...
-            "uint8",
-            "int8",
-            "int16",
-            "int32",
-            "int64",
-            "bool",
-        )
-    },
+@with_supported_dtypes(
+    {"2.0.1 and below": "float"},
```
Comment (vedpatwardhan) [inline ivy/functional/backends/torch/experimental/losses.py:107]: The error message in [this](https://github.com/unifyai/ivy/actions/runs/6301635175/job/17107481998?pr=26000#step:3:30698) test failure is coming from here, basically it should be `("float",)` instead of `"float"`

**Review by vedpatwardhan — CHANGES_REQUESTED** [review]
Just a minor change, thanks @ShreyanshBardia @HaiderSultanArc 😄

File: ivy/functional/backends/torch/experimental/losses.py (hunk 8 above)
Comment (ShreyanshBardia) [reply ivy/functional/backends/torch/experimental/losses.py]: updated it, but it looks like the tests are not running now due to `dill` not being available. Thanks!

**Comment by HaiderSultanArc** [comment]
@ShreyanshBardia I think you should run the tests locally and see if they pass 🤔 if `dill` missing error appears just do `pip install dill`

**Comment by ShreyanshBardia** [comment]
Hi @HaiderSultanArc, like I mentioned the tests were passing for me in local earlier also (now as well), it was the CI which was throwing error. Also reference to `dill` being missing was also with respect to `CI`.
